### ISSUE TITLE
[question\rfc] meson: environment variable usage rework

### ISSUE THREAD

**Issue by TheQwertiest** [conan-io/conan#7026]
Current `meson` integration uses the following approach:
- Generate environment via `AutoToolsBuildEnvironment`.
- Use env_vars (which sets up `CXXFLAGS`, `LDFLAGS` and etc) from this environment when executing all `ninja`/`meson` commands.

There are two issues(?) with such approach:
1. All dependency information (include paths, library search paths, defines and etc) is passed via global environment variables (e.g. `CPPFLAGS`), which means they apply to all the code in the package. Which result in a lot of extra paths to search, e.g. a case when a recipe has a lot of deps, but each library in this recipe only has a single dep. This also results in conflicts sometime: e.g. dep's headers are included with `-I`, but the project tries to include them with `-isystem` which might or might not work, depending on compiler and warning flags.
2. `meson` greatly discourages the use of environment variables and has explicit command line arguments to specify most of the flags instead. Current situation results in a weird situations, e.g. `fPIC` specified both in env vars and via command line. `meson` also might ignore some flags from env vars altogether.

Both problems can be fixed by not using `environment_append` in [this code](https://github.com/conan-io/conan/blob/02d9775c1c8dcd5580984ebf9a4fb638cbc1fd7a/conans/client/build/meson.py#L171) and by:

Problem 1.: using deps in `meson` through `pkg-config` generator (which is the recommended `mesonic` way).
Problem 2.: passing the arguments manually to `meson` during configuration.

Since this will be a breaking change, the easiest way would be to add a flag to `conans.Meson` ctor, e.g `Meson(generate_env=True)`, so by default the behaviour would remain the same, but the user could opt-in for the *clean* approach.

Note: this does approach does not aim to disable the whole env vars usage (e.g. from profiles), just the ones that are formed dynamically from `conan` settings and recipe deps.

I'm willing to try my hand at such PR, if such approach is approved.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
I agree it is worth improving this. 
I would vote for approaching this directly in the new ``toolchain()`` approach, ongoing work here: https://github.com/conan-io/conan/pull/5919 

This is the approach we are most likely to promote as default for Conan 2.0, and we are hopefully releasing it experimentally for CMake in next 1.26.

We would need some expertise in Meson for doing this the ``mesonic`` way, if you or more users of Meson could help, maybe we could try to do a first iteration of the Meson ``toolchain()`` in 1.27

**Comment by TheQwertiest** [comment]
I'm not really familiar with CMake, so please correct me if I've misread\misunderstood smth: from what I've seen from #5919 and #5737 , the idea is to generate some `toolchain` cmake file that contains various information about flags and compiler (which remotely resembles `conan` profiles) and pass it via a proepr cmake argument\env variable.

Toolchain concept in `meson` is a bit different from this *toolchain* and is more similar to `conan` profiles. The full proper invocation of `meson` configuration step for cross build would look smth like this:

`meson setup $BUILDDIR $CMD_ARGS --native-file=$NATIVE_FILE --cross-file=$CROSS_FILE`

where
- `$CMD_ARGS` - arguments that are not specific to build\host (e.g. `backend`, `unity-build`, `buildtype`)
- `$NATIVE_FILE` and `$CROSS_FILE` - so called `machine file`, files in the specific format that describe the build and host/target environment respectively (https://mesonbuild.com/Machine-files.html).

`meson` build is considered a `cross` build internally only when `$CROSS_FILE` is specified.
`$NATIVE_FILE` and `$CROSS_FILE` can be used separately (i.e. `$NATIVE_FILE` can be used in native build and is also optional in cross build).
Machine files also allow layering, so the user can additional files if they need to.

PR that started working on `cross-file` implementation: https://github.com/conan-io/conan/pull/6410
Related issue: https://github.com/conan-io/conan/issues/4529

**Comment by TheQwertiest** [comment]
The end result should look smth like this:
```py
class MachineFile:
  options = dict(); # parsed MachineFile, which is just a glorified ini file
  name = '';
  
  def dump(self, output):
    config = configparser.ConfigParser()
    outpath = Path(output)
    outpath.mkdir(parents=True)
    config.read_dict(options)
    with open(outpath/self.name, 'w') as f:
      config.write(f)

class MachineFiles:
  native_files = []
  cross_files = []
  
  def dump(self, output):
    if (native_files):
      outpath = Path(output)\'native'
      outpath.mkdir(parents=True)
      for f in native_files:
        f.dump(outpath)
    if (cross_files):
      outpath = Path(output)\'cross'
      outpath.mkdir(parents=True)
      for f in cross_files:
        f.dump(outpath)

def toolchain(self):
  # Generates cross file `if tools.cross_building() == true` or if `profile:host` is present
  # Generates native file `if tools.cross_building() == false` or if `profile:build` is present
  mcs = generate_default_machine_files(self)
  return mcs
```

**Comment by TheQwertiest** [comment]
And the recipe would look smth like this:
```py

class Pkg(ConanFile):
    name = 'pkg'
    version = '1.0'
    source_folder = '...'
    build_path = '...'

    def toolchain(self):
        mcs = generate_default_machine_files(self)
        mcs.cross_files += generate_machine_file(self, 'mah_cross'
                                                 { 'binaries': 
                                                   {'b2': '/usr/local/mypath/b2'} 
                                                 })
        return mcs

    def get_configured_meson(self):
        meson = MesonNew(self)
        # `meson.configure` should contain a check to prevent invocation on already configured folder
        meson.configure(source_folder=self.source_folder,
                        build_folder=self.build_path,
                        machine_files=meson.get_machine_files()
                        additional_args=[-Darg=1,...])
        return meson;

    def build(self):
        meson = self.get_configured_meson()
        meson.build()
        meson.test()

    def package(self):
        meson = self.get_configured_meson()
        meson.install()
```

**Comment by memsharded** [comment]
yes, exactly, when we call it ``toolchain`` we don't refer strictly to the cmake toolchain. In this case, what you are calling ``def generate_machine_files(self):`` would be exactly the thing to implement in the ``toolchain()`` method.

The idea is that the ``toolchain()``, run at ``conan install`` time, generates all the files that are necessary to run the build system natively, as a developer, without really needing to call it via conan to achieve the same result.

**Comment by TheQwertiest** [comment]
Amended both examples. Does that align properly with the *toolchain* idea?
How far is https://github.com/conan-io/conan/pull/5919 ? Can I use it as a basis? Or should I wait till it's implemented (and merged) completely before starting to work on `meson` implementation?

**Comment by memsharded** [comment]
#5919 is looking good so far, but surely still will change. I would suggest working on it once it is merged to the develop branch, no need to be released, but at least in develop, yes, to avoid having to repeat work. I am aiming to having it merged end of this week.

**Comment by TheQwertiest** [comment]
@memsharded , a follow up question (since I'm starting the work on the draft PR): what is the proper way to access env vars from build and host profile respectively?

[EDIT] I've looked around source code a bit and it seems it's impossible to do so. Hence a feature request: #7091

**Comment by TheQwertiest** [comment]
Note: Separate access for env vars is needed to generate both native and cross machine files (because meson might need to use compiler from build environment).

**Comment by SSE4** [comment]
most of work is done in #8147, #8111 and #7662. all configuration properties are in `native file` or `cross file`. need to check what else is needed.

**Comment by memsharded** [comment]
Hi @SSE4 Please check if https://github.com/conan-io/conan/pull/8147 finally implements this and if this can be closed.

**Comment by SSE4** [comment]
all dependencies are now located via `pkg-config`
the information like cflags is now passed via cross file or native file
we no longer pass things like fPIC or Visual Studio runtime via cflags, but use meson built-in properties for that purpose instead
I believe the issue could be closed

**Comment by memsharded** [comment]
Ok, so considered this implemtented, and will be released in Conan 1.33
